src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -2048,6 +2048,78 @@ abstract class RefChecks extends Transform {
+            def check(p: Tree, selSym: Symbol): Unit = {
+              val traverser = new Traverser {
+                // names absolved of shadowing because it is a "current" parameter (of a case class, etc)
+                var absolved: List[Name] = Nil
+                override def traverse(t: Tree): Unit = t match {
+                  case Apply(_, args) =>
+                    treeInfo.dissectApplied(t).core.tpe match {
+                      case MethodType(ps, _) =>
+                        foreach2(ps, args) { (p, arg) =>
+                          absolved ::= p.name
+                          try traverse(arg)
+                          finally absolved = absolved.tail
+                        }
+                      case _ => super.traverse(t)
+                    }
+                  case bind @ Bind(name, _) =>
+                    def richLocation(sym: Symbol): String = sym.ownsString match {
+                      case ""   => val n = sym.pos.line; if (n > 0) s"$sym at line $n" else sym.fullLocationString
+                      case owns => s"$sym in $owns"
+                    }
+                    for (shade <- bind.getAndRemoveAttachment[PatShadowAttachment]) {
+                      val shadowed = shade.shadowed
+                      if (!absolved.contains(name) && !bind.symbol.hasTransOwner(shadowed.accessedOrSelf) && checkShadowed(shadowed)(selSym))
+                        refchecksWarning(bind.pos, s"Name $name is already introduced in an enclosing scope as ${richLocation(shadowed)}. Did you intend to match it using backquoted `$name`?", WarningCategory.OtherShadowing)
+
+                    }
+                  case _ => super.traverse(t)
+                }
+              }
+              traverser.traverse(p)
+            }
+            // check the patterns for unfriendly shadowing, patvars bearing PatShadowAttachment
+            def checkShadowingPatternVars(): Unit =
+              selector match {
+                // or for `(x, y) match case p(x, y) =>` check the corresponding args
+                case treeInfo.Applied(Select(core, nme.apply), _, tupleArgs :: Nil) if isTupleSymbol(core.symbol.companion) =>
+                  for (cdef <- cases)
+                    cdef.pat match {
+                      case Apply(_, args) if sameLength(args, tupleArgs) =>
+                        foreach2(args, tupleArgs)((arg, sel) => check(arg, notNull(sel.symbol)))
+                      case UnApply(_, args) if sameLength(args, tupleArgs) =>
+                        foreach2(args, tupleArgs)((arg, sel) => check(arg, notNull(sel.symbol)))
+                      case p => check(p, selectorSymbol)
+                    }
+                // or for `tp.dealiased match case tp =>` check `tp` if it looks like a derived value, but not `this.x` or `x.asInstanceOf`
+                case treeInfo.Applied(Select(core, _), _, _) if core.symbol != null && !core.symbol.isThisSym && selector.tpe <:< core.tpe.widen =>
```

Comment:
What do you think about generalizing it to "shadowed symbol shows up in scrutinee"?

https://github.com/som-snytt/scala/compare/issue/11850...lrytz:scala:pr8806?expand=1

Reply:
I thought about it but was too lazy to experiment. I considered that `n + 1 match case n` should be the same as `1 + n`. Requiring a conforming type means it doesn't take arbitrary `f(x) match case x`. Generalizing is easier to explain and less fussy. OK, I will try it out.

Reply:
OK, I already forgot that I intended to follow up with an experiment.